mbbsd/register.c

```diff
@@ -563,6 +599,108 @@ do_register_captcha()
 // New Registration (Phase 1: Create Account)
 /////////////////////////////////////////////////////////////////////////////
 
+static int
+new_register_email_verify(char *email)
+{
+    clear();
+    vs_hdr("E-Mail �{��");
+    move(1, 0);
+    outs("�z�n, �����n�D��U�ɶi�� E-Mail �{��:\n"
+	 "  �п�J�z�� E-Mail , �ڭ̷|�H�o�t���{�ҽX���H�󵹱z\n"
+	 "  �����Ш�п�J�{�ҽX, �Y�i�i���U\n"
+	 "  ��: ���������� yahoo, kimo���K�O�� E-Mail\n"
+	 "\n"
+	 "**********************************************************\n"
+	 "* �Y�L�[������Ш�l��U�����ˬd�O�_�Q��@�U���H(SPAM)�F,*\n"
+	 "* �t�~�Y��J��o�ͻ{�ҽX��~�Х�T�{��J�O�_���̫�@��   *\n"
+	 "* ���쪺�{�ҫH�A�Y�u�����M����ЦA����@�� E-Mail.       *\n"
+	 "**********************************************************\n");
+
+    // Get a valid email from user.
+    int err;
+    int tries = 0;
+    do {
+	if (++tries > 10) {
+	    return REGISTER_ERR_CANCEL;
+	}
+
+	// Use "-" for userid, regmaild rejects empty userid, which we don't
+	// really have a userid yet.
+	err = register_email_input("-", email);
+	switch (err) {
+	    case REGISTER_OK:
+		if (strcasecmp(email, "x") != 0)
+		    break;
+		// Use input is "x".
+		err = REGISTER_ERR_INVALID_EMAIL;
+		// fallthrough
+	    case REGISTER_ERR_INVALID_EMAIL:
+		move(15, 0); clrtobot();
+		move(17, 0);
+		outs("��w�� E-Mail �����T�C�i��A��J���O�K�O��Email�A\n");
+		outs("�δ����ϥΪ̥H�� E-Mail �{�ҫ�Q�����C\n\n");
+		pressanykey();
+		continue;
+
+	    case REGISTER_ERR_TOO_MANY_ACCOUNTS:
+		move(15, 0); clrtobot();
+		move(17, 0);
+		outs("��w�� E-Mail �w��U�L�h�b��, �ШϥΨ�L E-Mail.\n");
+		pressanykey();
+		continue;
+
+	    default:
+		return err;
+	}
+    } while (err != REGISTER_OK);
+
+    // Send and check regcode.
+    char inregcode[14] = {0}, regcode[50];
+    char buf[80];
+    makeregcode(regcode);
+
+    tries = 0;
+    int num_sent = 0;
+    bool send_code = true;
+    while (1) {
+	if (++tries > 10) {
+	    vmsg("��~���ƹL�h, �еy��A��");
+	    return REGISTER_ERR_CANCEL;
+	}
+	if (send_code) {
+	    tries++;  // Expensive operation.
+	    email_regcode(regcode, email);
+	    send_code = false;
+	    num_sent++;
+	    snprintf(buf, sizeof(buf),
+		    ANSI_COLOR(1;31) " (�� %d ��)" ANSI_RESET, num_sent);
+	}
+
+	move(15, 0); clrtobot();
+	outs("�{�ҽX�w�H�e��:");
+	if (num_sent > 1) outs(buf);
+	outs("\n");
+	outs(ANSI_COLOR(1;33) "  ");
+	outs(email);
+	outs(ANSI_RESET "\n");
+	outs("�{�ҽX�`�@����Q�T�X, �S���ťզr��.\n");
+	outs("�Y�L�[�S����{�ҽX�i��J x ���s�H�e.\n");
+
+	getdata(20, 0, "�п�J�{�ҽX�G",
+		inregcode, sizeof(inregcode), DOECHO);
+	if (strcasecmp(inregcode, "x") == 0) {
+	    send_code = true;
+	    continue;
+	}
+	if (strcasecmp(inregcode, regcode) != 0) {
```

Comment:
Maybe also handle empty string separately? Like not count it as an attempt?

Reply:
In the rest of registration flow, empty input is treated as an error, or a failed attempt.

Reply:
Ok. Consistency is good.